mypy/test/data/check-inference.test

```diff
@@ -690,6 +690,60 @@ l = lb # E: Incompatible types in assignment (expression has type List[bool], va
 [builtins fixtures/for.py]
 
 
+-- Generic function inference with unions
+-- --------------------------------------
+
+
+[case testUnionInference]
+from typing import TypeVar, Union, List
+T = TypeVar('T')
+U = TypeVar('U')
+def f(x: Union[T, int], y: T) -> T: pass
+f(1, 'a')() # E: "str" not callable
+f('a', 1)() # E: "object" not callable
+
+def g(x: Union[T, List[T]]) -> List[T]: pass
+def h(x: List[str]) -> None: pass
+g('a')() # E: List[str] not callable
+g(['a']) # E: Argument 1 to "g" has incompatible type List[str]; expected "Union[None, List[None]]"
+h(g(['a']))
+
+def i(x: Union[List[T], List[U]], y: List[T], z: List[U]) -> None: pass
+a = [1]
+b = ['b']
+i(a, a, b)
+i(b, a, b)
+i(a, b, b) # E: Argument 1 to "i" has incompatible type List[int]; expected "Union[List[str], List[str]]"
```

Comment:
The union type comes from the user-declared type of `i`. Then it becomes redundant when substituting for `T` and `U`. I'll make type substitution simplify unions when possible (this is good in other situations too).